### PR TITLE
Add rudimentary test coverage for drake-visualizer.

### LINKED ISSUE #2767: drake-visualizer needs test coverage in CI
https://github.com/RobotLocomotion/drake/issues/2731 should have been caught immediately by CI -- drake-visualizer simply failed to run on homebrew mac.  We do not currently have any test coverage for running the drake-visualizer.

### PR DESCRIPTION
**Pull request by david-german-tri**
This test confirms that drake-visualizer will start up.  Since the Linux CI instances do not have Xvfb yet, and Windows does not have Director, this test only covers OS X.

Resolves #2767.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/2917)

### CODE REVIEW

**Comment by david-german-tri** [comment]
@drake-jenkins-bot mac-clang-experimental please

**Comment by david-german-tri** [comment]
@drake-jenkins-bot mac-clang-experimental please

**Comment by david-german-tri** [comment]
`mac-clang-experimental` did not actually run the new test.  Thus far, I am unable to reproduce this behavior locally, but I'll continue poking around.

**Comment by jamiesnape** [comment]
@drake-jenkins-bot retest this please

**Comment by david-german-tri** [comment]
@drake-jenkins-bot mac-clang-experimental please

(just in case)

**Comment by david-german-tri** [comment]
@drake-jenkins-bot mac-clang-experimental please

**Comment by jamiesnape** [comment]
> Since the Linux CI instances do not have Xvfb yet

They do now.

**Comment by david-german-tri** [comment]
Oh, I see.  https://github.com/RobotLocomotion/drake-ci/blame/master/ctest_driver_script.cmake does not propagate down `WITH_DIRECTOR`.  @jamiesnape and @mwoehlke-kitware, would you recommend updating the `ctest_driver_script.cmake` to propagate `WITH_DIRECTOR` into the drake `CMakeCache`, or using some other technique to determine whether Director is present?

**Comment by david-german-tri** [comment]
@BetsyMcPhail also

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2917#-:-KNbPoStBT0yN0vP_vW_:398557965)*

**Comment by jamiesnape** [comment]
Adding it to the CI script now.

**Comment by jamiesnape** [comment]
Done.

**Comment by david-german-tri** [comment]
Thanks, @jamiesnape! (Commit was https://github.com/RobotLocomotion/drake-ci/commit/2f50705d8c9fc03987d9b766eeade93b29b035b5)

**Comment by david-german-tri** [comment]
@drake-jenkins-bot: retest this please

**Comment by david-german-tri** [comment]
@drake-jenkins-bot: mac-clang-experimental please

**Comment by david-german-tri** [comment]
@drake-jenkins-bot: mac-clang-experimental please

**Comment by david-german-tri** [comment]
Interestingly, this approach doesn't work on the Mac CI builders.  Possibly a side-effect of vSphere.

```
93: _RegisterApplication(), FAILED TO establish the default connection to the WindowServer, _CGSDefaultConnection() is NULL.
```

https://drake-jenkins.csail.mit.edu/job/mac-clang-experimental/70/console

**Comment by jamiesnape** [comment]
Have you tried locally using SSH? I would assume that would fail too.

**Comment by david-german-tri** [comment]
> Have you tried locally using SSH? I would assume that would fail too.

Indeed!  It passes over SSH if the user running the test also happens to be logged into the desktop, but fails in this manner otherwise.  So the question becomes, can we cause an SSH session to spawn a desktop session?  I'm concerned the answer to that is probably no, but I'll try to confirm.

**Comment by jamiesnape** [comment]
There are possibilities, but it will need some reconfiguration of the VM image first.

**Comment by jamiesnape** [comment]
In the meantime, so we get some test coverage, I would propose limiting the test to just Linux and then opening another PR for other platforms after that is merged.

**Comment by david-german-tri** [comment]
@drake-jenkins-bot: mac-clang-experimental please

**Comment by david-german-tri** [comment]
> In the meantime, so we get some test coverage, I would propose limiting the test to just Linux and then opening another PR for other platforms after that is merged.

Done!  @jamiesnape and @mwoehlke-kitware for feature review.

**Comment by david-german-tri** [comment]
@drake-jenkins-bot: mac-clang-experimental please

File: drake/regtests/visualizer_startup_test.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Tests that DrakeVisualizerApp can be initialized without crashing.
```
Comment (jamiesnape) [inline drake/regtests/visualizer_startup_test.py]: `#!/usr/bin/env python` at top of file.

Comment (jamiesnape) [inline drake/regtests/visualizer_startup_test.py]: Format comment as docstring.

**Comment by jamiesnape** [comment]
:lgtm:

---

Reviewed 2 of 4 files at r1, 1 of 2 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2917#-:-KNloqsVX1kKVYotWJjs:1417088721)*

**Comment by david-german-tri** [comment]
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/regtests/visualizer_startup_test.py, line 1 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2917#-KNXgnGHcxaF8VTI5lLs-r5-1:-KNlpwC1iG0LFmKjQL66:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/4c56d2d4771dcb0c350243f5acc10b2727d712c2/drake/regtests/visualizer_startup_test.py#L1)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary>
> Format comment as docstring.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2917)*

**Comment by jamiesnape** [comment]
@drake-jenkins-bot retest this please

**Comment by jamiesnape** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2917#-:-KNmbDqyvf97bPfPNJ0m:1417088721)*

**Comment by jamiesnape** [comment]
BTW @david-german-tri The OS X images have been updated as discussed F2F.

**Comment by david-german-tri** [comment]
+@jwnimmer-tri for platform review

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2917#-:-KNn7ZfcyPvyM9Ey-9yV:-58081829)*

**Comment by jwnimmer-tri** [comment]
All of my comments are BTWs (lowercased to avoid the reviewable.io bug), so I'll pre-mark this :lgtm:.

---

Reviewed 1 of 4 files at r1, 1 of 2 files at r4, 1 of 1 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[CMakeLists.txt, line 157 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2917#-KNnKJTL6SBmvSZGEwFE:-KNnKJTNStlHR2q8iLQY:1055328231) ([raw file](https://github.com/robotlocomotion/drake/blob/168e7b823270ed9a3ce25e46f30502ff8ae60a31/CMakeLists.txt#L157)):*
> ```CMake
>   -DWITH_PYTHON_3=${WITH_PYTHON_3}
>   -DDISABLE_MATLAB=${DISABLE_MATLAB}
>   -DWITH_DIRECTOR=${WITH_DIRECTOR}
> ```

btw This is going to get old fast, isn't it?

---

*[drake/regtests/CMakeLists.txt, line 13 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2917#-KNnSznsVx90TeKtddSJ:-KNnSzntqjCGTl4WwRFK:-421610306) ([raw file](https://github.com/robotlocomotion/drake/blob/168e7b823270ed9a3ce25e46f30502ff8ae60a31/drake/regtests/CMakeLists.txt#L13)):*
> ```CMake
> else()
>   set(hasDisplay FALSE)
> endif()
> ```

btw Consider moving this hunk (and its comment block) within the `if(WITH_DIRECTOR)`?  Not sure why it would be most readable _before_ the matlab testing clauses, while director stuff is after?

---

*[drake/regtests/CMakeLists.txt, line 22 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2917#-KNnTDVH4FiIhaau63tF:-KNnTDVH4FiIhaau63tG:1895459046) ([raw file](https://github.com/robotlocomotion/drake/blob/168e7b823270ed9a3ce25e46f30502ff8ae60a31/drake/regtests/CMakeLists.txt#L22)):*
> ```CMake
>   # A test that starts up drake-visualizer.
>   set(visualizerStartupTestCommand
>       ${CMAKE_INSTALL_PREFIX}/bin/directorPython ${CMAKE_CURRENT_SOURCE_DIR}/visualizer_startup_test.py)
> ```

btw Consider keeping to less than 80 columns.

---

*[drake/regtests/CMakeLists.txt, line 28 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2917#-KNnLHuJaJFon5bZTmXl:-KNnLHuKL3zmisdXQlKD:1895459046) ([raw file](https://github.com/robotlocomotion/drake/blob/168e7b823270ed9a3ce25e46f30502ff8ae60a31/drake/regtests/CMakeLists.txt#L28)):*
> ```CMake
>   elseif(xvfb-run)
>     add_test(NAME visualizer_startup_test_xvfb
>              COMMAND xvfb-run -s "-screen 0 1024x768x24" ${visualizerStartupTestCommand})
> ```

btw Consider keeping to less than 80 columns.

---

*[drake/regtests/visualizer_startup_test.py, line 10 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2917#-KNnTO2bkR9stfRrgDyD:-KNnTO2bkR9stfRrgDyE:-794709258) ([raw file](https://github.com/robotlocomotion/drake/blob/168e7b823270ed9a3ce25e46f30502ff8ae60a31/drake/regtests/visualizer_startup_test.py#L10)):*
> ```Python
> app = vis.DrakeVisualizerApp()
> app.mainWindow.show()
> app.quit()
> ```

btw Are there any threads launched by director?  If so, consider adding a sleep before the quit, to let things crash in the background?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2917#-:-KNnTarmQ5KoqbY0EerP:-1141197869)*

**Comment by mwoehlke-kitware** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[CMakeLists.txt, line 157 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2917#-KNnKJTL6SBmvSZGEwFE:-KNqn37zTS8vZSjTKB1Q:1851271105) ([raw file](https://github.com/robotlocomotion/drake/blob/168e7b823270ed9a3ce25e46f30502ff8ae60a31/CMakeLists.txt#L157)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> btw This is going to get old fast, isn't it?
</details>
It may be possible to arrange for all `WITH_` to be propagated... however this will result in a lot of warning spam about specified-but-unused variables. (We're already getting some of that from e.g. the Java flags being passed to everyone.)

A better solution might be to configure a file with all of these and have drake-proper import it.

For obvious reasons, I'd prefer to wait until after the refactor (#2896) for either solution...

---

*[drake/regtests/visualizer_startup_test.py, line 10 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2917#-KNnTO2bkR9stfRrgDyD:-KNqoZinzQax2w4Ci-SA:-1795384662) ([raw file](https://github.com/robotlocomotion/drake/blob/168e7b823270ed9a3ce25e46f30502ff8ae60a31/drake/regtests/visualizer_startup_test.py#L10)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> btw Are there any threads launched by director?  If so, consider adding a sleep before the quit, to let things crash in the background?
</details>
Note: ~~sleep~~[`app.processEvents`](http://doc.qt.io/qt-4.8/qcoreapplication.html#processEvents-2)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2917)*

**Comment by jwnimmer-tri** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[CMakeLists.txt, line 157 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2917#-KNnKJTL6SBmvSZGEwFE:-KNqrLPjc9qvkuvGqjPt:55714964) ([raw file](https://github.com/robotlocomotion/drake/blob/168e7b823270ed9a3ce25e46f30502ff8ae60a31/CMakeLists.txt#L157)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary>
> It may be possible to arrange for all `WITH_` to be propagated... however this will result in a lot of warning spam about specified-but-unused variables. (We're already getting some of that from e.g. the Java flags being passed to everyone.)
> 
> A better solution might be to configure a file with all of these and have drake-proper import it.
> 
> For obvious reasons, I'd prefer to wait until after the refactor (#2896) for either solution...
</details>
Sure, happy to wait.  However, perhaps its a defect if a `WITH_` option is never examined within drake?  It seems like every single one of them should effect the Drake build in some way.  In other words, if `WITH_FOO` is set, Drake should bomb out if it's search for _foo_ doesn't find anything.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2917)*

**Comment by mwoehlke-kitware** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[CMakeLists.txt, line 157 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2917#-KNnKJTL6SBmvSZGEwFE:-KNqu7PfKyjpAynZoSHr:-1851321560) ([raw file](https://github.com/robotlocomotion/drake/blob/168e7b823270ed9a3ce25e46f30502ff8ae60a31/CMakeLists.txt#L157)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Sure, happy to wait.  However, perhaps its a defect if a `WITH_` option is never examined within drake?  It seems like every single one of them should effect the Drake build in some way.  In other words, if `WITH_FOO` is set, Drake should bomb out if it's search for _foo_ doesn't find anything.
</details>
Usually, in drake itself, you just look for the package, and use it if it is found. In a sense, needing to propagate *any* of these could be considered a bug.

That said, propagating them in order to verify that enabled packages *are* found seems like a reasonable use case, though I don't think I've seen that done before...

That said, the variables for required dependencies (e.g. Eigen, where the option is to use the version from the superbuild or the system version) should not need to be propagated in any case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2917)*

**Comment by david-german-tri** [comment]
Review status: 1 of 4 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---

*[drake/regtests/CMakeLists.txt, line 13 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2917#-KNnSznsVx90TeKtddSJ:-KNr208z0pd0nPRTRuEN:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/168e7b823270ed9a3ce25e46f30502ff8ae60a31/drake/regtests/CMakeLists.txt#L13)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> btw Consider moving this hunk (and its comment block) within the `if(WITH_DIRECTOR)`?  Not sure why it would be most readable _before_ the matlab testing clauses, while director stuff is after?
</details>
Done.

---

*[drake/regtests/CMakeLists.txt, line 22 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2917#-KNnTDVH4FiIhaau63tF:-KNr289ixuyFHLnC90m7:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/168e7b823270ed9a3ce25e46f30502ff8ae60a31/drake/regtests/CMakeLists.txt#L22)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> btw Consider keeping to less than 80 columns.
</details>
Done.

---

*[drake/regtests/CMakeLists.txt, line 28 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2917#-KNnLHuJaJFon5bZTmXl:-KNr28SIeS81VH_VquR-:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/168e7b823270ed9a3ce25e46f30502ff8ae60a31/drake/regtests/CMakeLists.txt#L28)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> btw Consider keeping to less than 80 columns.
</details>
Done.

---

*[drake/regtests/visualizer_startup_test.py, line 10 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2917#-KNnTO2bkR9stfRrgDyD:-KNr68iqab1zYneWbJEu:395210162) ([raw file](https://github.com/robotlocomotion/drake/blob/168e7b823270ed9a3ce25e46f30502ff8ae60a31/drake/regtests/visualizer_startup_test.py#L10)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary>
> Note: ~~sleep~~[`app.processEvents`](http://doc.qt.io/qt-4.8/qcoreapplication.html#processEvents-2)
</details>
`app.processEvents` appears not to exist, according to my Python interpreter.  I'm not going to pretend to understand Qt at all (and I'd rather not spend time learning), so I'm just going to defend the existing approach by pointing out it's as thread-safe as Director's analogous test case:

https://github.com/RobotLocomotion/director/blob/master/src/python/tests/testDrakeVisualizer.py

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2917)*

**Comment by david-german-tri** [comment]
@drake-jenkins-bot mac-clang-experimental please

**Comment by jwnimmer-tri** [comment]
Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2917)*

**Comment by david-german-tri** [comment]
@jamiesnape Works great on Mac!  Thanks!

**Pull request merged**